enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,327 @@
+  information.  Can we do this? 
+
+## Summary
+
+There are a few key network connections that are important to the health of the cluster.
+We can create a CRD to store the the last success time for each point-to-point check.
+By watching for explicit failures and lack of updates, we can produce a picture of when a point-to-point check failed.
+This information could be correlated to e2e failures, alerts, events, logs, or other information feeds.
+
+## Motivation
+
+This attempts to identify failures of particular connections in a black-box manner that is independent of SDN, cloud,
+or cluster ingress.
+
+### Goals
+
+1. be independent of SDN, cloud, or cluster ingress
+2. Identify connection failure between each kas and each oas endpoint
+3. Identify connection failure between each kas and each etcd
+4. Identify connection failure between each kas and oas service IP
+5. Identify connection failure between each oas and each kas host IP
+6. Identify connection failure between each oas and each kas endpoint IP
+7. Identify connection failure between each oas and each etcd
+
+### Non-Goals
+
+1. DNS resolution, that would be distinct.
+
+## Proposal
+
+1. One instance of a new CR per point to point connection.
+2. Detailed status, probably some kind of latency information.
+3. It should be possible to clean up garbage after a couple hours of inactivity.
+4. Don't delete aggressively, on upgrades, we have different endpoints
+5. Create a binary that can take multiple destinations paired with instances to write to.
+6. Use that binary in containers included in each kas and oas (or other interesting) pod.
+
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+#### RelatedEndpoint Custom Resource
+
+```go
+package v1alpha1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConditionCheck
+type ConditionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ConditionCheckSpec   `json:"spec"`
+	Status ConditionCheckStatus `json:"status"`
+}
+
+type ConditionCheckSpec struct {
+	// SourcePod names the pod from which the condition will be checked
+	SourcePod string `json:"source_pod"`
+	// Endpoint being checked
+	Endpoint string `json:"endpoint"`
+	// Type of check performed on the endpoint
+	Type ConditionCheckType `json:"type"`
+}
+
+type ConditionCheckStatus struct {
+	// Log of check events
+	Log []LogEntry `json:"log"`
```

Comment:
separate successes and failures

Reply:
> separate successes and failures

There is a `Status` field in the log entry for indicating Success/Failure.

Reply:
.status.successes and .status.failures allows you to answer questions like: when was the last time this failed.  and "how long has this been good"

Reply:
Updated.